package.json

```diff
@@ -60,6 +60,7 @@
     "@rollup/plugin-node-resolve": "^15.0.1",
     "@testing-library/jest-dom": "^5.16.4",
     "@testing-library/react": "^12.1.5",
+    "@testing-library/user-event": "^14.5.2",
```

Comment:
Have you considered any alternatives instead of adding the new dependency?

Reply:
Good catch. This library conflicts with our test utils, so it is better to avoid it.

Related past thread: https://github.com/cloudscape-design/board-components/pull/235#discussion_r1195047786

Reply:
I've tried a few different options before using that library: 

1. fire a `change` event on the native input, but that would defeat the purpose of the test because `change` event isn't fired when the same file is re-uploaded, and here I would be forcing to fire which isn't true in a real browser.

2. Add an integ test with text on the page the counts how many times `onChange` was called, but that had some limitations: 
   - I need to add a file in the page so I can upload it and that can be tricky in CI environment.
   -  webdriver upload function is only supported when running Chrome or using a Selenium Grid with the Selenium standalone server: https://webdriver.io/blog/2019/06/25/file-upload/
   - the technique in the above example uses `setValue` and that actually triggers on change even if the same file is re-uploaded so the test becomes trivial.

3. copy this [block of code in user-event library](https://github.com/testing-library/user-event/blob/main/src/utility/upload.ts#L36-L55) that mimics the a real file input behaviour into our test code. 


So I was left with only 2 options: use `@testing-library/user-event` or copy it's `upload` function implementation, I went with the first option.

@just-boris I don't understand from that thread how this library conflicts with our test-utils, the point about to exposing this functionality to customers shouldn't conflict with customers using a popular library to do the same functionality.

I went with adding the library because I saw that it had other useful utilities that could improve other tests, e.g. [Clipboard](@testing-library/user-event) for [CopyToClipboard](https://github.com/cloudscape-design/components/blob/main/src/copy-to-clipboard/__tests__/copy-to-clipboard.test.tsx#L20)


I'm happy to disagree and commit, also let me know if there is a better way to test this other than the ways I tried

Reply:
I think writing an e2e test is the only option here. 

User-event works for testing the change events, but that library has other downsides (in their click events).

So I would prefer not seeing this library at all, than remembering a long list of allowed/not allowed methods

Reply:
@just-boris e2e test wouldn't work as I mentioned in `2.3` above because it actually calls `onChange` even if the same file is re-uploaded, that has to do with the fact that `<input type='file'>` does not allow setting a value programmatically to anything other than an empty string https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#notes

so when a e2e test framework sets a value to a file input it's also not the real user experience

Reply:
Discussed in DM. Let's try these options

1. `Simulate.change` from `react-dom/test-utils`
2. Override methods on the input element our own way
3. If both options are unfeasible, let's skip test coverage for this case, unfortunately

Reply:
tried both options: 
1. Simulate.change has the same behaviour as fireEvent 
2. I tried overriding the methods but seems like it's more complicated than the lines I've included above from user-event, other overrides are needed elsewhere, see [here](https://github.com/testing-library/user-event/blob/main/src/utils/dataTransfer/FileList.ts), [here](https://github.com/testing-library/user-event/blob/main/src/utils/edit/setFiles.ts#L28) and [here](https://github.com/testing-library/user-event/blob/main/src/event/dispatchEvent.ts#L8)
I've tried to only override what's needed for my test but it's not enough, based on the amount of overrides needed I think we're much diverting from a real input element so I just removed the test.

Reply:
Then we call it as "untestable", because testing against user-event does not represent real browser behavior

:(